sql/sql_base.cc

```diff
@@ -3800,6 +3800,31 @@ open_and_process_table(THD *thd, TABLE_LIST *tables, uint *counter, uint flags,
 
     thd->pop_internal_handler();
     safe_to_ignore_table= no_such_table_handler.safely_trapped_errors();
+
+#ifdef WITH_WSREP
+    if (WSREP(thd) && !thd->wsrep_applier && !error)
+    {
+      /*
+        Append a table level shared key for the referenced/foreign table.
+        This to avoid potential MDL conflicts with concurrent DDLs.
+      */
+      const int rcode = wsrep_thd_append_table_key(thd, tables->db.str,
+          tables->table_name.str, WSREP_SERVICE_KEY_SHARED);
```

Comment:
key type here should be WSREP_SERVICE_KEY_REFERENCE. Shared type is historically reserved for key "branches", i..e in the case of db.table.row UPDATE key three keys will be added to the key set:
1. db - SHARED
2. db.table - SHARED
3. db.table.row - UPDATE

WSREP_SERVICE_KEY_SHARED shouild not be used explicitly for adding keys.

Reply:
I have to disagree with that shared keys should not be used. Shared key has still a valid use in a case we want to create a dependency (not conflict) between two write sets for which the first one has a shared key and the second an exclusive key for a resource which both write sets will access.

See the table for dependency rules: https://github.com/codership/Engineering/wiki/Foreign_Key_Optimization_Combined#dependency-rules